### PR TITLE
A0-3256: Check for equivocations

### PR DESCRIPTION
**Pull request by lesniak43**
# Description

Detecting Aura equivocations, `panic!` when self-equivocating

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have added tests
- I have created new documentation

### CODE REVIEW

**Review by timorleph — APPROVED** [review]
Looks good, one question – is the `None` variant for `AccountId` only there because of runtime APIs not supporting this in the past? Can this be fixed with our new planned approach to this API, and if yes is there a task for that?

File: finality-aleph/src/sync/service.rs [hunk 1]
```diff
@@ -313,6 +313,15 @@ where
         }
     }
 
+    fn process_equivocation_proofs(&self, proofs: Vec<V::EquivocationProof>) {
+        for proof in proofs {
+            warn!(target: LOG_TARGET, "Equivocation detected: {proof}");
+            if proof.are_we_equivocating() {
+                panic!("We are equivocating, which is ILLEGAL - shutting down the node. This is probably caused by running two instances of the node with the same set of credentials. Make sure that you are running ONLY ONE instance of the node. If the problem persists, contact the Aleph Zero developers on Discord.");
```
Comment (kostekIV) [inline finality-aleph/src/sync/service.rs:320]: we really should have just a page with a contact info and just past a link here -.-

File: finality-aleph/src/sync/substrate/verification/cache.rs [hunk 2]
```diff
@@ -76,6 +76,7 @@ where
     H: Header,
 {
     cached_data: HashMap<SessionId, CachedData>,
+    pub equivocation_cache: HashMap<u64, (H, bool)>,
```
Comment (kostekIV) [inline finality-aleph/src/sync/substrate/verification/cache.rs:79]: does it need to be a pub?

File: finality-aleph/src/sync/service.rs (hunk 1 above)
Comment (lesniak43) [reply finality-aleph/src/sync/service.rs:320]: I guess that we might create a task for this? Although I've talked with @DamianStraszak and he says this should be fine.

Comment (DamianStraszak) [reply finality-aleph/src/sync/service.rs:320]: I guess it's fine, it's informative and practical. I don't know if we have any page with such contact information.

Comment (kostekIV) [reply finality-aleph/src/sync/service.rs:320]: A task would be also about creating such a page. For this pr here its fine for now

**Comment by lesniak43** [comment]
@timorleph yes, and no - I'll remember to create the task as soon as we have the related epic (so that we don't have to prioritize the task right now)

File: finality-aleph/src/sync/substrate/verification/cache.rs (hunk 2 above)
Comment (lesniak43) [reply finality-aleph/src/sync/substrate/verification/cache.rs]: ofc not, refactored the code slightly

**Review by kostekIV — APPROVED** [review]

**Pull request merged**